src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlDelegatedTransaction.cs

```diff
@@ -331,27 +328,36 @@ public void Rollback(SysTx.SinglePhaseEnlistment enlistment)
                     connection.CleanupConnectionOnTransactionCompletion(_atomicTransaction);
                     enlistment.Aborted();
                 }
+                catch (System.OutOfMemoryException e)
+                {
+                    usersConnection.Abort(e);
+                    throw;
+                }
+                catch (System.StackOverflowException e)
+                {
+                    usersConnection.Abort(e);
+                    throw;
+                }
+                catch (System.Threading.ThreadAbortException e)
+                {
+                    usersConnection.Abort(e);
+                    throw;
+                }
 #if DEBUG
                 finally
                 {
                     tdsReliabilitySection.Stop();
                 }
 #endif //DEBUG
             }
-            catch (System.OutOfMemoryException e)
+            catch (ObjectDisposedException)
```

Comment:
I remember you talking about ObjectDisposedException. However an ObjectDisposedException being found is an indicator that we are re-using a disposed object and it shouldn't be handled in a try/catch

https://docs.microsoft.com/en-us/dotnet/api/system.objectdisposedexception?view=netframework-4.8

`In most cases, this exception results from developer error. Instead of handling the error in a try/catch block, you should correct the error, typically by reinstantiating the object.
`

Reply:
I could change the logic to NOT throw object disposed exception, but that would also apply to other places where we DO want to throw this exception, that's why I went this way.

This is very specific to Rollback/Abort, since when there's a race condition, we see different types of transaction terminations at server side:
1. Internal Abort
2. Rollback - Triggered by driver when we receive notifications to Rollback the transaction. (Calls are intermittent and may/may not happen due to application design. DTC might have already parallelly aborted the tx.)

Also commented in this method above and can be seen in the code of `TransactionEnded`, we don't really care if this `_connection` is already aborted or null.

Rollback on a non-active connection should not fail.

Reply:
i want to know why are we getting the ObjectDispisedException in the first place. What is the object and type in question.
We shouldn't be getting this exception at all.
I agree that rollback shouldn't throw an exception. My point is that we shouldn't be catching the ObjectDisposedException but removing the reason this exception occurred. Occurrence of this exception is more often than not due to a bug that should be chased down.

Reply:
We get this exception from [here](https://github.com/dotnet/SqlClient/pull/495/files/2f59a6f4b00fd53b2d2322271e6b1f3f9681bfd7#diff-2a7d4b53e8840c0897e356a092430483R525) because TransactionEnded got executed before Rollback, in race condition.

![image](https://user-images.githubusercontent.com/13396919/78312238-f989cd00-7507-11ea-94ee-6949e2dc2e16.png)

The callbacks are triggered externally, so we have no control over that.

Reply:
We got this exception from [here](https://github.com/dotnet/SqlClient/pull/495/files/2f59a6f4b00fd53b2d2322271e6b1f3f9681bfd7#diff-2a7d4b53e8840c0897e356a092430483R525) because TransactionEnded (triggerd by DTC > SysTx) got executed before Rollback (triggered by SysTx), in race condition and it set `_connection` to null before it was acquired for Rollback to execute.

![image](https://user-images.githubusercontent.com/13396919/78312479-c5fb7280-7508-11ea-96c8-e2bdedda851f.png)

Since these callbacks are external, we cannot control their order of execution.

Reply:
OK, looks like that the GetValidConnection() was only called to check if the user is using a Disposed connection to call this API.
We should be able to track that the transaction ended call was triggered due to the callback, even though we cannot control the timing of the callback trigger since it is Sys.Transaction triggered, which is DTC triggered. 
This affirms that we shouldn't be swallowing the ObjectDisposedException because that was probably added for Api correctness.